Messages+ Extension

SharkPool-SP opened the pull request:
New Powerful Message Blocks that work with **Vanilla Blocks** and **More Events**!
![image](https://github.com/TurboWarp/extensions/assets/139097378/bf0698c8-3367-47c1-bb68-56cbdffd01b7)

SharkPool-SP committed:
Create Messages-Plus.js

SharkPool-SP committed:
Create Messages-Plus.svg

veggiecan0419 commented:
👍'd ya own PR :trollface: @SharkPool-SP?

SharkPool-SP commented:
> 👍'd ya own PR :trollface: @SharkPool-SP?

need to start the trend somehow :trollface:

veggiecan0419 commented:
Regarding the formatting error, you can open a codespace, open a terminal, and then you can run `npm run format`.
Hope that helps!
(as all your PRs seem to have a miserable x)

![x](https://github.com/TurboWarp/extensions/assets/137440114/2edbb6bd-6fb1-461b-9407-f6040d538894)

Congrats on all the extensions tho!

SharkPool-SP commented:
> Regarding the formatting error, you can open a codespace, open a terminal, and then you can run `npm run format`. Hope that helps! (as all your PRs seem to have a miserable x)
> 
> ![x](https://private-user-images.githubusercontent.com/137440114/321859708-2edbb6bd-6fb1-461b-9407-f6040d538894.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI4OTg0OTEsIm5iZiI6MTcxMjg5ODE5MSwicGF0aCI6Ii8xMzc0NDAxMTQvMzIxODU5NzA4LTJlZGJiNmJkLTZmYjEtNDYxYi05NDA3LWY2MDQwZDUzODg5NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxMlQwNTAzMTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00M2UyZDIwNzMyYWUwOWM3M2EwMmI0NTliNmNiMTdmODMxMjE5MGRlMjA0Mzk1ZDUyZGQ0ZmY5ZmU5MWZmNTFlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.2CWZUGNbH66rtcUR5zGIO7qAJPuWENLk1h6TEBt8pbA)
> 
> Congrats on all the extensions tho!

Thanks! (I have no life :trollface:) 
I could format it but Id rather let Garbo do it. I also dont have any idea what a valid codespace is so yeah...

veggiecan0419 commented:
And, could I place these #1390  in here, or should I make a new extension?

SharkPool-SP commented:
Yes and no, I would want you to modify your stuff using the functions in my extension. Since you don't need to make edits to the vm

SharkPool-SP commented:
Talk to me on discord, I can explain further

veggiecan0419 commented:
> Talk to me on discord, I can explain further

@SharkPool-SP I don't have discord. 

We could talk about it in https://github.com/veggiecan0419/extensions/issues/7 instead

SharkPool-SP committed:
Update Messages-Plus.js

SharkPool-SP committed:
Update Messages-Plus.js

SharkPool-SP committed:
Update Messages-Plus.js

SharkPool-SP committed:
Update Messages-Plus.js

SharkPool-SP committed:
Messages-Plus -- Faster _getMessageHats function

SharkPool-SP committed:
Messages-Plus -- I forgot 2 brackets

SharkPool-SP committed:
Update Messages-Plus.js

SharkPool-SP commented:
update <img width="536" alt="Screenshot 2024-08-30 at 11 54 50 PM" src="https://github.com/user-attachments/assets/6aea6452-2ac4-471c-9fe1-1b12b406d7b4">

SharkPool-SP committed:
Messages-Plus -- Add Overlap Option

SharkPool-SP committed:
Messages-Plus.js -- bugfix

SharkPool-SP commented:
!format

DangoCat committed:
[Automated] Format code

GarboMuffin committed:
Merge branch 'master' into patch-35

GarboMuffin committed:
remove thing that is never used

GarboMuffin committed:
returning null or undefined in the same function is ugly

GarboMuffin committed:
use the right received data

GarboMuffin committed:
concat is not in-place

Brackets-Coder commented:
Just a quick code skim, haven't tested or critically evaluated yet

If this was going to be added to the gallery (which I presume it is) it needs a `// License:` header and a line in the JSON

Brackets-Coder commented on extensions/SharkPool/Messages-Plus.js:
```diff
@@ -0,0 +1,652 @@
+      for (const thread of threads) thread[kMessageName] = name;
+    }
+    return threads;
+  };
```
Assuming this is supposed to sync native broadcasts with Messages+. Is there a scenario where a user might want the two separated?

Brackets-Coder replied:
Nevermind, just realized this is an odd question after reading the description

SharkPool-SP committed:
Add license and code style improvements

GarboMuffin commented on extensions/SharkPool/Messages-Plus.js:
```diff
@@ -0,0 +1,667 @@
+  const kMessageName = Symbol("kMessageName"); // May be defined on a Thread as a string
+  const kReceivedData = Symbol("kReceivedData"); // May be defined on a Thread as any Scratch-compatible value
+
+  // TODO: _all_ is not actually a reserved value
```
if there's a todo then what's the actionable next step here to do

SharkPool-SP replied:
that was your comment

GarboMuffin replied:
cool

GarboMuffin commented on extensions/SharkPool/Messages-Plus.js:
```diff
@@ -0,0 +1,667 @@
+      let waiting = false;
+      for (const ID of blockIds) {
+        const thread = runtime.threads.find((thread) => thread.topBlock === ID);
+        if (thread && thread.messageFlag === undefined) {
```
you mostly use thread[kSymbolSomething], which is good

some spots where you somewhat don't, like here

and probably the same w/ justReported elsewhere

maybe they shoudl all be symbols

SharkPool-SP replied:
done

SharkPool-SP committed:
improvements

DangoCat committed:
[Automated] Format code

SharkPool-SP committed:
better banner

SharkPool-SP committed:
Add SharkPool/Messages-Plus to extensions list

SharkPool-SP committed:
fix ratio

SharkPool-SP committed:
Update description for Messages-Plus.js

SharkPool-SP committed:
Add documentation URI to Messages-Plus.js

SharkPool-SP committed:
add documentation

Added detailed documentation for the Messages extension, including block overviews and usage examples.

Brackets-Coder commented on docs/SharkPool/Messages-Plus.md:
````diff
@@ -0,0 +1,167 @@
+```
+
+Broadcasts the specified message to the selected sprites while sending a value to the receiving scripts,
+then returns an array containing the responses from every receiver.
````
Remove the line break here. It looks odd in my browser

<img width="740" height="458" alt="Image" src="https://github.com/user-attachments/assets/2e8cd7bb-d79e-4034-a7cf-53301318e8ec" />

Brackets-Coder commented on extensions/SharkPool/Messages-Plus.js:
```diff
@@ -0,0 +1,671 @@
+            shouldRestartExistingThreads: true,
+            text: Scratch.translate("when any broadcast is received"),
+          },
+          {
+            opcode: "messageName",
+            blockType: Scratch.BlockType.REPORTER,
+            extensions: ["colours_event"],
+            allowDropAnywhere: true,
+            disableMonitor: true,
+            text: Scratch.translate("broadcast name"),
+          },
```
This returns the wrong value when broadcasting multiple messages with the array block

<img width="1840" height="1196" alt="Image" src="https://github.com/user-attachments/assets/bf59cce9-ce4f-4644-8902-568d30215bf3" />

It returns lowercase for non-array messages

SharkPool-SP replied:
Ah, not a bug: 
<img width="1067" height="752" alt="image" src="https://github.com/user-attachments/assets/426ec706-ae33-4b27-91c9-db6f6ca4d6de" />

message2 is not defined in the menu.

Brackets-Coder reviewed (dismissed):
There was something odd about the script overlap and script restart blocks. I don't know if I wasn't understanding or if they weren't working properly, but it seems like script overlap wasn't creating multiple instances of the broadcast.

<img width="1840" height="1196" alt="Image" src="https://github.com/user-attachments/assets/c910edba-7c51-49bf-aa1e-b9aa225a638d" />

This is a short review because I can't get in a big one at the moment

Brackets-Coder commented on docs/SharkPool/Messages-Plus.md:
Docs look okay

Brackets-Coder commented on extensions/SharkPool/Messages-Plus.js:
```diff
@@ -0,0 +1,671 @@
+              WAIT: { type: Scratch.ArgumentType.STRING, menu: "SHOULD_WAIT" },
+            },
+          },
+          {
+            blockType: Scratch.BlockType.XML,
+            xml: `
+              <block type="SPmessagePlus_broadcastTarget"><value name="BROADCAST_OPTION"><shadow type="event_broadcast_menu"></shadow></value><value name="TARGET"><shadow type="SPmessagePlus_menu_TARGETS"></shadow></value><value name="DATA"><shadow type="text"></shadow></value><value name="SHOULD_WAIT"><shadow type="SPmessagePlus_menu_SHOULD_WAIT"></shadow></value></block>
+              <block type="SPmessagePlus_broadcastDataTarget"><value name="BROADCAST_OPTION"><shadow type="event_broadcast_menu"></shadow></value><value name="TARGET"><shadow type="SPmessagePlus_menu_TARGETS"></shadow></value><value name="DATA"><shadow type="text"></shadow></value><value name="SHOULD_WAIT"><shadow type="SPmessagePlus_menu_SHOULD_WAIT"></shadow></value></block>
+            `,
+          },
```
Could you explain how XML blocks work? Also, why are two XML blocks defined in one chunk of JSON? Wouldn't it be better or easier to read if they were separated?

SharkPool-SP replied:
They basically display blocks in the palette using their XML form. This allows you to do cool things like having preset scripts in the palette, or in my case, injecting the broadcast message menu into blocks.

2 XML chunks are defined in the same spot to avoid lag with XML generation. Which iirc it can happen. Its also just to keep organized for which block goes after another.

Brackets-Coder commented on extensions/SharkPool/Messages-Plus.js:
```diff
@@ -0,0 +1,671 @@
+    }
+
+    // Block Funcs
+    messageName(args, util) {
+      if (!Object.prototype.hasOwnProperty.call(util.thread, kMessageName)) {
+        return "";
+      }
+
+      const name = util.thread[kMessageName];
+      // The name stored on the thread is toUpperCase(), so lookup the user-facing name
+      const variable = util.runtime
+        .getTargetForStage()
+        .lookupBroadcastByInputValue(name);
+      if (variable) return variable.name;
+      return name; // this is a dynamic message
+    }
```
toUpperCase may have been part of the incorrect user-facing name when returning the message name from an array broadcast

SharkPool-SP replied:
toUpperCase is required since messages internally (when called for the 'when I receive' blocks to run) are converted to upper case.

Ill add it to the docs that dynamic messages (onces that arent pre-defined) will be converted to uppercase

SharkPool-SP committed:
Fix line break in docs

SharkPool-SP committed:
mention dynamic message quirks in the docs

GarboMuffin commented on extensions/SharkPool/Messages-Plus.js:
```diff
@@ -0,0 +1,671 @@
+      }
+
+      util.stackFrame.initialized = false;
+      return JSON.stringify(util.stackFrame.responses);
```
if your extension is not named "JSON" then having blocks that return JSON is a major smell

this block is now unusable without people also knowing how to use some JSON extension (not always obvious)

also, you've now lost the ability for a broadcast to return NaN or Infinity

GarboMuffin replied:
yes, many other extensions do, at least they do it in a context where it's inconceivable that you'll get something that isn't JSON serializable

SharkPool-SP replied:
So should I make it an option to return the first response as a string?

Not sure what you want me to do here.

SharkPool-SP replied:
Fixed:
<img width="373" height="499" alt="image" src="https://github.com/user-attachments/assets/1b6eef31-3c59-4105-b02d-208151adbd9d" />

Its off by default

GarboMuffin commented on extensions/SharkPool/Messages-Plus.js:
```diff
@@ -0,0 +1,671 @@
+      return waiting;
+    }
+
+    receivers(args) {
```
how do you imagine someone is going to use this?

GarboMuffin replied:
like what are they going to do with it? especially with it not differentiating clones at all

SharkPool-SP replied:
Debugging could be one for sure.

SharkPool-SP replied:
Or perhaps sprites loaded via asset manager that have recieve blocks

GarboMuffin commented on extensions/SharkPool/Messages-Plus.js:
```diff
@@ -0,0 +1,671 @@
+      }
+
+      for (const thread of threads) {
+        if (thread[kResponseData]) {
```
falsey return value gets deleted

<img width="340" height="275" alt="Image" src="https://github.com/user-attachments/assets/6c9e4284-37c4-4e6b-83ad-97060fa5ec8c" />

SharkPool-SP replied:
fixed

GarboMuffin commented on extensions/SharkPool/Messages-Plus.js:
```diff
@@ -0,0 +1,671 @@
+      const broadcastName = Cast.toString(args.BROADCAST_OPTION).toUpperCase();
+      const now = Date.now();
+      let waiting = false;
+      for (const thread of runtime.threads) {
```
If someone makes a bunch of "forever { if received < > { ... } }" for some reason, CPU time will be O(n^2). But aside from that: You stamp the time even if the thread did not receive anything

consider this one: `b` gets set to "detected". which makes sense

<img width="284" height="521" alt="Image" src="https://github.com/user-attachments/assets/fc3110fb-a38a-40bf-ad02-078a30133b62" />

if i make no change to either of those scripts and add an additional new script polling for receiving for some reason, `b` will now get set to "missed"

<img width="542" height="521" alt="Image" src="https://github.com/user-attachments/assets/aa272637-9405-4190-bfbd-41b7622011e4" />

SharkPool-SP replied:
Fixed

GarboMuffin commented on extensions/SharkPool/Messages-Plus.js:
```diff
@@ -0,0 +1,671 @@
+        }
+        if (
+          thread[kMessageName] === broadcastName &&
+          thread[kReceiveTime] >= now
```
same-millisecond calls will be allowed through. it's weird to allow project behavior to be accidentally dependent on code executing in the same millisecond if they check this multiple times in succession and expect the result to not change

it also results in situations like this: what should the value of `i` be after running the loop in this scenario? currently, you get something mostly random based on how many times the block gets to run in 1 ms

<img width="388" height="383" alt="Image" src="https://github.com/user-attachments/assets/238ab018-cec9-4099-adce-e873923d4bf7" />

SharkPool-SP replied:
Fixed

SharkPool-SP committed:
some fixes

SharkPool-SP committed:
update docs

SharkPool-SP committed:
Fix '<is () received?>'

DangoCat committed:
[Automated] Format code

GarboMuffin committed:
fix default response value for non-multi when nothing got started

GarboMuffin committed:
fix docs

DangoCat committed:
[Automated] Format code

SharkPool-SP committed:
smarter <is () received?>

DangoCat committed:
[Automated] Format code

GarboMuffin committed:
i perceive it to be used as an alternative to "when i receive", not as an additional thing. so don't require that event hats for the boolean to work

GarboMuffin reviewed (approved):
review the last commit then merge as desired if you have no complaints

GarboMuffin commented:
ok i might just override the 2 reviewer requirement depending on vibes tonight

GarboMuffin committed:
update docs
